File: Programmers/키패드 누르기.py
import math


def solution(numbers, hand):
    answer = ''
    hand_position = [[3, 0], [3, 2]]
    for i in range(len(numbers)):
        if numbers[i] == 1 or numbers[i] == 4 or numbers[i] == 7:
            answer += "L"
            hand_position[0] = cal_position(numbers[i])
        elif numbers[i] == 3 or numbers[i] == 6 or numbers[i] == 9:
            answer += "R"
            hand_position[1] = cal_position(numbers[i])
        else:
            current_pos = cal_position(numbers[i])
            left_dis = abs(current_pos[0] - hand_position[0][0]) + abs(current_pos[1] - hand_position[0][1])
            right_dis = abs(current_pos[0] - hand_position[1][0]) + abs(current_pos[1] - hand_position[1][1])

            if left_dis > right_dis:
                answer += "R"
                hand_position[1] = current_pos
            elif left_dis < right_dis:
                answer += "L"
                hand_position[0] = current_pos
            else:
                if hand == "left":
                    answer += "L"
                    hand_position[0] = current_pos
                else:
                    answer += "R"
                    hand_position[1] = current_pos

    return answer


def cal_position(num):
    if num == 0:
        return [3, 1]
    else:
        temp = num - 1
        row = temp // 3
        col = temp % 3
        return [row, col]

File: Programmers JS/키패드 누르기.js
function solution(numbers, hand) {
  var answer = "";
  let test = Array.from({ length: 5 }, (v, i) => i);
  console.log(test);
  let left = "*";
  let right = "#";
  for (let num of numbers) {
    if (num === 1 || num === 4 || num === 7) {
      answer += "L";
      left = num;
    } else if (num === 3 || num === 6 || num === 9) {
      answer += "R";
      right = num;
    } else {
      let leftDistance = calculateDistance(left, num);
      let rightDistance = calculateDistance(right, num);
      if (leftDistance < rightDistance) {
        answer += "L";
        left = num;
      } else if (leftDistance > rightDistance) {
        answer += "R";
        right = num;
      } else {
        if (hand === "right") {
          answer += "R";
          right = num;
        } else {
          answer += "L";
          left = num;
        }
      }
    }
  }
  return answer;
}

function calculateDistance(current, target) {
  let prev = getLocation(current);
  let post = getLocation(target);

  let x = Math.abs(prev[0] - post[0]);
  let y = Math.abs(prev[1] - post[1]);
  return x + y;
}

function getLocation(number) {
  switch (number) {
    case "*":
      return [3, 0];
    case "#":
      return [3, 2];
    case 0:
      return [3, 1];
    default:
      return [parseInt((number - 1) / 3), (number - 1) % 3];
  }
}
